smartsim/_core/mli/infrastructure/control/requestdispatcher.py

```diff
@@ -0,0 +1,525 @@
+
+    def release(self) -> None:
+        self._lock.release()
+
+    def __enter__(self) -> None:
+        self.acquire()
+
+    def __exit__(
+        self,
+        exc_type: t.Optional[t.Type[BaseException]],
+        exc_val: t.Optional[BaseException],
+        exc_tb: t.Optional[TracebackType],
+    ) -> None:
+        self.release()
+
+
+class BatchQueue(Queue[InferenceRequest]):
+    def __init__(self, batch_timeout: float, batch_size: int, model_key: str) -> None:
+        """Queue used to store inference requests waiting to be batched and
+        sent to Worker Managers.
+        :param batch_timeout: Time in seconds that has to be waited before flushing a
+        non-full queue. The time of the firt item put is 0 seconds.
+        :param batch_size: Total capacity of the queue.
+        :param model_key: Key of the model which needs to be executed on the queued
+        requests
+        """
+        super().__init__(maxsize=batch_size)
+        self._batch_timeout = batch_timeout
+        """Time in seconds that has to be waited before flushing a non-full queue.
+        The time of the firt item put is 0 seconds."""
+        self._batch_size = batch_size
+        """Total capacity of the queue."""
+        self._first_put: t.Optional[float] = None
+        """Time at which the first item was put on the queue"""
+        self._disposable = False
+        """Whether the queue will not be used again and can be deleted.
+        A disposable queue is always full."""
+        self._model_key = model_key
+        """Key of the model which needs to be executed on the queued requets"""
+        self._flush_lock = RLock()
+        """Lock used to make sure only one process can flush the queue (unused now)"""
+        self._id = str(uuid.uuid4())
+        """Id of queue"""
+
+    @property
+    def queue_id(self) -> str:
+        """ID of this queue"""
+        return self._id
+
+    def acquire(self, blocking: bool = True, timeout: float = -1) -> t.Optional[bool]:
+        """Acquire queue lock to flush
+        :param blocking: whether to block on lock acquisition
+        :param timeout: Time to wait if blocking, before raising exception
+        """
+        return self._flush_lock.acquire(blocking=blocking, timeout=timeout)
+
+    def release(self) -> None:
+        """Release queue lock"""
+        self._flush_lock.release()
+
+    def __enter__(self) -> None:
+        self.acquire()
+
+    def __exit__(
+        self,
+        exc_type: t.Optional[t.Type[BaseException]],
+        exc_val: t.Optional[BaseException],
+        exc_tb: t.Optional[TracebackType],
+    ) -> None:
+        self.release()
+
+    @property
+    def model_key(self) -> str:
+        """Key of the model which needs to be run on the queued requests"""
+        return self._model_key
+
+    def put(
+        self,
+        item: InferenceRequest,
+        block: bool = False,
+        timeout: t.Optional[float] = 0.0,
```

Comment:
Do you think we'll have a need to change the timeout per-request? If not, we might be able to clean up the `put` method by setting the timeout globally in the queue initializer.

Reply:
I think this function gets called by `put_nowait` and I may not be able to redefine its signature, I'll give it a try